File: Longest Sub-Array with Sum K - GFG/longest-sub-array-with-sum-k.py
#User function Template for python3

class Solution:
    def lenOfLongSubarr (self, arr, n, k) : 
        #Complete the function
        
        dict = {}
        maxi =0
        sum =0
        
        for i in range(len(arr)):
            
            sum += arr[i]
            
            if(not sum in dict): dict[sum] = i
            
            
            if sum == k: maxi = max(i+1,maxi)
            if sum-k in dict :
                maxi = max(i-dict[sum-k],maxi)
                
        return maxi
                
    



#{ 
 # Driver Code Starts
#Initial Template for Python 3


for _ in range(0,int(input())):
    
    n, k = map(int , input().split())
    arr = list(map(int,input().strip().split()))
    ob = Solution()
    print(ob.lenOfLongSubarr(arr, n, k))




# } Driver Code Ends

File: Longest Sub-Array with Sum K - GFG/longest-sub-array-with-sum-k.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution{
    public:
    int lenOfLongSubarr(int nums[],  int n, int k) 
    { 
        // Complete the function
        
        int sum=0;
        map<int,int>m;
        int maxi = 0;
        
        
        for(int i=0; i<n;i++){
            
            sum = sum + nums[i];
            if(m.find(sum) == m.end()) m[sum] = i; 
            
            if(sum == k) maxi = max(maxi,i+1);
            
            else if(m.find(sum-k) != m.end()){
                maxi = max(maxi,i-m[sum-k]);
            }
        }

        return maxi;
    } 

};

//{ Driver Code Starts.

int main() {
	//code
	
	int t;cin>>t;
	while(t--)
	{
	    int n, k;
	    cin>> n >> k;
	    int a[n];
	    
	    for(int i=0;i<n;i++)
	        cin>>a[i];
	   Solution ob;
	   cout << ob.lenOfLongSubarr(a, n , k)<< endl;
	    
	}
	
	return 0;
}
// } Driver Code Ends